src/iterative_ensemble_smoother/experimental_sies.py

```diff
@@ -0,0 +1,315 @@
+        *,
+        seed: Optional[int] = None,
+    ):
+        self.iteration = 1
+        self.rng = np.random.default_rng(seed)
+        self.X = param_ensemble
+        self.d = observation_values
+        self.C_dd = observation_errors
+        self.A = (self.X - self.X.mean(axis=1, keepdims=True)) / np.sqrt(
+            self.X.shape[1] - 1
+        )
+
+        if self.C_dd.ndim == 2:
+            self.C_dd_cholesky = sp.linalg.cholesky(
+                self.C_dd,
+                lower=True,
+                overwrite_a=False,
+                check_finite=True,
+            )
+        else:
+            self.C_dd_cholesky = np.sqrt(self.C_dd)
+
+        # Equation (14)
+        self.D = self.d.reshape(-1, 1) + sample_mvnormal(
+            C_dd_cholesky=self.C_dd_cholesky, rng=self.rng, size=self.X.shape[1]
+        )
+
+        self.W = np.zeros(shape=(self.X.shape[1], self.X.shape[1]))
+        self.X_i = self.X
+
+    def newton(self, Y, step_length=0.5):
+        """Implementation of Algorithm 1."""
+        g_X = Y.copy()
+
+        # Get shapes. Same notation as used in the paper.
+        N = Y.shape[1]  # Ensemble members
+        n = self.X.shape[0]  # Parameters (inputs)
+        m = self.C_dd.shape[0]  # Responses (outputs)
+
+        # Line 4 in Algorithm 1
+        Y = (g_X - g_X.mean(axis=1, keepdims=True)) / np.sqrt(N - 1)
+
+        # Line 5
+        Omega = self.W.copy()
+        Omega -= Omega.mean(axis=1, keepdims=True)
+        Omega /= np.sqrt(N - 1)
+        Omega.flat[:: Omega.shape[0] + 1] += 1  # Add identity in place
+
+        # Line 6
+        if n < N - 1:
+            # There are fewer parameters than realizations. This means that the
+            # system of equations is overdetermined, and we must solve a least
+            # squares problem.
+
+            # An alternative approach to producing A_i would be keeping the
+            # returned value from the previous Newton iteration (call it X_i),
+            # then computing:
+            # A_i = (self.X_i - self.X_i.mean(axis=1, keepdims=True)) / np.sqrt(N - 1)
+            A_i = self.A @ Omega
+            S = sp.linalg.solve(
+                Omega.T, np.linalg.multi_dot([Y, sp.linalg.pinv(A_i), A_i]).T
+            ).T
+        else:
+            # The system is underdetermined
+            S = sp.linalg.solve(Omega.T, Y.T).T
+
+        # Line 7
+        H = S @ self.W + self.D - g_X
+
+        # Line 8
+        # Some asserts to remind us what the shapes of the matrices are
+        assert self.W.shape == (N, N)
+        assert S.shape == (m, N)
+        assert self.C_dd.shape in [(m, m), (m,)]
+        assert H.shape == (m, N)
+        self.W = inversion_exact(
+            W=self.W,
+            step_length=step_length,
+            S=S,
+            C_dd=self.C_dd,
+            H=H,
+            C_dd_cholesky=self.C_dd_cholesky,
+        )
+
+        # Line 9
+        return self.X + self.X @ self.W / (np.sqrt(N - 1))
```

Comment:
Do we need a parenthesis around `np.sqrt(N-1)`?

Reply:
nope!